akka-actor/src/main/scala/akka/actor/CoordinatedShutdown.scala

```diff
@@ -354,16 +351,120 @@ final class CoordinatedShutdown private[akka] (
     system: ExtendedActorSystem,
     phases: Map[String, CoordinatedShutdown.Phase])
     extends Extension {
-  import CoordinatedShutdown.Reason
-  import CoordinatedShutdown.UnknownReason
+  import CoordinatedShutdown.{ Reason, UnknownReason }
 
   /** INTERNAL API */
   private[akka] val log = Logging(system, getClass)
   private val knownPhases = phases.keySet ++ phases.values.flatMap(_.dependsOn)
 
   /** INTERNAL API */
   private[akka] val orderedPhases = CoordinatedShutdown.topologicalSort(phases)
-  private val tasks = new ConcurrentHashMap[String, Vector[(String, () => Future[Done])]]
+
+  private trait PhaseDefinition {
```

Comment:
isn't the separate trait for PhaseDefinition overkill?

Reply:
I don't think so. The inner `loop` function in `run` is already fairly complex, since its responsibilities include popping a phase off the stack, determining if it's enabled, running it and possibly racing against the timeout. Better to let the phase handle its own running rather than forcing `loop` to worry about that too.

Reply:
This trait is internal though, and has only one concrete subtype, so we can just skip it and talk about the concrete type directly instead. That doesn't change anything with loop or run. I think you can just rename the `StrictPhaseDefinition` to `PhaseDefinition` and drop this.

Reply:
All true, though the original idea was to aid maintainability by removing the temptation for anyone working outside `tasks` to copy/modify a `PhaseDefinition`. In particular, refactoring the concrete case class out to this level of indentation would require pulling a whole bunch of currently encapsulated/private stuff out of `tasks`, which I'd rather not do if it isn't necessary.

Reply:
Can't the concrete the type stay where it is then, and just reference that from where PhaseDefinition is used now?

Reply:
Not without leaking abstraction. The case class parameter list mentions `TaskDefinition`, which is internal to `tasks`, for example.

Reply:
Ok, let's leave it be for now then.